File: src/queenscoach/transit.py
"""Domain layer: joins the realtime feeds to the static schedule and resolves
the human-friendly queries the tools accept ("9", "Blue Line", "CTC").
"""

from __future__ import annotations

import math
import re
from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Literal

from .realtime import ServiceAlert, TripUpdate, VehiclePosition
from .static_gtfs import Mode, Route, Schedule, Stop

_METERS_PER_SECOND_TO_MPH = 2.2369363
_EARTH_RADIUS_METERS = 6_371_000

VehicleMode = Literal["bus", "train", "unknown"]


@dataclass(frozen=True, slots=True)
class ResolvedRouteRef:
    route_id: str
    name: str
    long_name: str
    mode: Mode


@dataclass(frozen=True, slots=True)
class NextStop:
    stop_id: str
    name: str | None
    arrival_time: str
    minutes_away: int


@dataclass(frozen=True, slots=True)
class VehicleView:
    #: Vehicle number shown on the bus or train.
    vehicle: str
    mode: VehicleMode
    route: ResolvedRouteRef | None
    headsign: str | None
    trip_id: str | None
    latitude: float
    longitude: float
    bearing_degrees: float | None
    speed_mph: float | None
    occupancy: str | None
    reported_at: str | None
    position_age_seconds: int | None
    next_stop: NextStop | None


@dataclass(frozen=True, slots=True)
class ArrivalView:
    route: ResolvedRouteRef | None
    headsign: str | None
    vehicle: str | None
    trip_id: str | None
    arrival_time: str
    minutes_away: int
    scheduled_arrival_time: str | None
    #: Positive means running late. ``None`` when the feed gives no schedule.
    schedule_deviation_minutes: int | None
    vehicle_position: tuple[float, float] | None


@dataclass(frozen=True, slots=True)
class TransitSnapshot:
    schedule: Schedule
    vehicles: Sequence[VehiclePosition]
    trip_updates: Sequence[TripUpdate]
    #: Unix time, in seconds, that this snapshot describes.
    now: float


@dataclass(frozen=True, slots=True)
class _Upcoming:
    stop_id: str
    arrival_time: int


def normalize(text: str) -> str:
    """Casefold and collapse whitespace, for comparing user queries to feed text."""
    return re.sub(r"\s+", " ", text.strip().lower())


def natural_key(text: str) -> tuple[tuple[int, int, str], ...]:
    """Sort key that orders embedded numbers numerically, so 9 precedes 29 and 501."""
    return tuple(
        (1, int(part), "") if part.isdigit() else (0, 0, part)
        for part in re.split(r"(\d+)", text)
        if part
    )


def iso_time(epoch_seconds: float) -> str:
    """Format Unix seconds as an ISO 8601 UTC timestamp, e.g. ``2026-01-09T14:03:00.000Z``."""
    stamp = datetime.fromtimestamp(epoch_seconds, tz=UTC)
    return stamp.isoformat(timespec="milliseconds").replace("+00:00", "Z")


def to_iso_time(epoch_seconds: float | None) -> str | None:
    """Like :func:`iso_time`, but passes ``None`` through."""
    return None if epoch_seconds is None else iso_time(epoch_seconds)


def _minutes_between(epoch_seconds: float, now: float) -> int:
    return round((epoch_seconds - now) / 60)


def distance_meters(from_lat: float, from_lon: float, to_lat: float, to_lon: float) -> float:
    """Great-circle distance in meters."""
    delta_lat = math.radians(to_lat - from_lat)
    delta_lon = math.radians(to_lon - from_lon)
    a = (
        math.sin(delta_lat / 2) ** 2
        + math.cos(math.radians(from_lat))
        * math.cos(math.radians(to_lat))
        * math.sin(delta_lon / 2) ** 2
    )
    return 2 * _EARTH_RADIUS_METERS * math.asin(min(1.0, math.sqrt(a)))


def _to_route_ref(route: Route | None) -> ResolvedRouteRef | None:
    if route is None:
        return None
    return ResolvedRouteRef(
        route_id=route.route_id,
        name=route.short_name,
        long_name=route.long_name,
        mode=route.mode,
    )


def find_routes(schedule: Schedule, query: str) -> list[Route]:
    """Resolve a user-supplied route query against the schedule.

    Matching is exact-first (``route_id``, then ``route_short_name``) so a query
    of "5" cannot be captured by "Route 501"; substring matching on the long
    name is the last resort.
    """
    needle = normalize(query)
    if not needle:
        return []
    routes = list(schedule.routes.values())

    for exact in (
        [route for route in routes if normalize(route.route_id) == needle],
        [route for route in routes if normalize(route.short_name) == needle],
        [route for route in routes if normalize(route.long_name) == needle],
    ):
        if exact:
            return exact

    return [
        route
        for route in routes
        if needle in normalize(route.long_name) or needle in normalize(route.short_name)
    ]


def find_stops(schedule: Schedule, query: str, limit: int = 10) -> list[Stop]:
    """Resolve a user-supplied stop query.

    Exact ``stop_id``/``stop_code`` first, then exact name, then a substring
    match ranked by name length so the closest match to the query leads.
    """
    needle = normalize(query)
    if not needle:
        return []

    direct = schedule.stops.get(query.strip())
    if direct is not None:
        return [direct]

    stops = list(schedule.stops.values())
    by_code = [stop for stop in stops if stop.code is not None and normalize(stop.code) == needle]
    if by_code:
        return by_code[:limit]

    by_name = [stop for stop in stops if normalize(stop.name) == needle]
    if by_name:
        return by_name[:limit]

    matches = [stop for stop in stops if needle in normalize(stop.name)]
    matches.sort(key=lambda stop: (len(stop.name), stop.name))
    return matches[:limit]


def matches_vehicle_query(vehicle: VehiclePosition, query: str) -> bool:
    """Match a vehicle query against its label, descriptor id, or entity id."""
    needle = normalize(query)
    if not needle:
        return False
    candidates = (vehicle.vehicle_label, vehicle.vehicle_id, vehicle.entity_id)
    return any(value is not None and normalize(value) == needle for value in candidates)


def _build_next_stop_index(trip_updates: Iterable[TripUpdate], now: float) -> dict[str, _Upcoming]:
    """Index the soonest upcoming stop per trip, from the TripUpdates feed.

    This is the only trustworthy source of a vehicle's next stop: the
    VehiclePositions feed's own ``stop_id`` and ``current_stop_sequence`` do not
    match the published schedule (see the note in :mod:`queenscoach.realtime`).
    """
    index: dict[str, _Upcoming] = {}
    for update in trip_updates:
        if update.trip_id is None:
            continue
        best: _Upcoming | None = None
        for stop_time in update.stop_time_updates:
            stop_id = stop_time.stop_id
            arrival_time = stop_time.arrival_time
            if stop_id is None or arrival_time is None or arrival_time < now:
                continue
            if best is None or arrival_time < best.arrival_time:
                best = _Upcoming(stop_id=stop_id, arrival_time=arrival_time)
        if best is not None:
            index[update.trip_id] = best
    return index


def to_vehicle_view(
    vehicle: VehiclePosition,
    snapshot: TransitSnapshot,
    next_stop_index: dict[str, _Upcoming],
) -> VehicleView:
    """Build the enriched, rider-facing view of a single vehicle."""
    schedule = snapshot.schedule
    trip = None if vehicle.trip_id is None else schedule.trips.get(vehicle.trip_id)
    # Prefer the route the vehicle reports; fall back to the one on its trip.
    route_id = vehicle.route_id or (trip.route_id if trip is not None else None)
    route = None if route_id is None else schedule.routes.get(route_id)

    upcoming = None if vehicle.trip_id is None else next_stop_index.get(vehicle.trip_id)
    next_stop = (
        None
        if upcoming is None
        else NextStop(
            stop_id=upcoming.stop_id,
            name=(
                schedule.stops[upcoming.stop_id].name
                if upcoming.stop_id in schedule.stops
                else None
            ),
            arrival_time=iso_time(upcoming.arrival_time),
            minutes_away=_minutes_between(upcoming.arrival_time, snapshot.now),
        )
    )

    speed = vehicle.speed_meters_per_second
    return VehicleView(
        vehicle=vehicle.vehicle_label or vehicle.vehicle_id or vehicle.entity_id,
        mode=route.mode if route is not None else "unknown",
        route=_to_route_ref(route),
        headsign=trip.headsign if trip is not None else None,
        trip_id=vehicle.trip_id,
        latitude=vehicle.latitude,
        longitude=vehicle.longitude,
        bearing_degrees=vehicle.bearing_degrees,
        speed_mph=None if speed is None else round(speed * _METERS_PER_SECOND_TO_MPH, 1),
        occupancy=vehicle.occupancy_status,
        reported_at=to_iso_time(vehicle.timestamp),
        position_age_seconds=(
            None if vehicle.timestamp is None else max(0, round(snapshot.now - vehicle.timestamp))
        ),
        next_stop=next_stop,
    )


def to_vehicle_views(
    vehicles: Iterable[VehiclePosition], snapshot: TransitSnapshot
) -> list[VehicleView]:
    """Build vehicle views for many vehicles, sharing one next-stop index."""
    index = _build_next_stop_index(snapshot.trip_updates, snapshot.now)
    return [to_vehicle_view(vehicle, snapshot, index) for vehicle in vehicles]


def arrivals_at_stop(
    stop: Stop,
    snapshot: TransitSnapshot,
    *,
    route_ids: frozenset[str] | None = None,
    mode: Mode | None = None,
) -> list[ArrivalView]:
    """Predicted arrivals at one stop, soonest first."""
    schedule = snapshot.schedule
    now = snapshot.now

    vehicle_by_trip = {
        vehicle.trip_id: vehicle for vehicle in snapshot.vehicles if vehicle.trip_id is not None
    }

    arrivals: list[ArrivalView] = []
    for update in snapshot.trip_updates:
        trip = None if update.trip_id is None else schedule.trips.get(update.trip_id)
        route_id = update.route_id or (trip.route_id if trip is not None else None)
        route = None if route_id is None else schedule.routes.get(route_id)

        if route_ids is not None and (route_id is None or route_id not in route_ids):
            continue
        if mode is not None and (route is None or route.mode != mode):
            continue

        for stop_time in update.stop_time_updates:
            if stop_time.stop_id != stop.stop_id:
                continue
            if stop_time.schedule_relationship == "SKIPPED":
                continue
            arrival_time = stop_time.arrival_time or stop_time.departure_time
            if arrival_time is None or arrival_time < now:
                continue

            vehicle = None if update.trip_id is None else vehicle_by_trip.get(update.trip_id)
            scheduled = stop_time.scheduled_arrival_time

            arrivals.append(
                ArrivalView(
                    route=_to_route_ref(route),
                    headsign=trip.headsign if trip is not None else None,
                    vehicle=(vehicle.vehicle_label if vehicle is not None else None)
                    or update.vehicle_label,
                    trip_id=update.trip_id,
                    arrival_time=iso_time(arrival_time),
                    minutes_away=_minutes_between(arrival_time, now),
                    scheduled_arrival_time=to_iso_time(scheduled),
                    # The feed omits `delay`, so deviation is derived from the two stamps.
                    schedule_deviation_minutes=(
                        None if scheduled is None else round((arrival_time - scheduled) / 60)
                    ),
                    vehicle_position=(
                        None if vehicle is None else (vehicle.latitude, vehicle.longitude)
                    ),
                )
            )

    arrivals.sort(key=lambda arrival: arrival.arrival_time)
    return arrivals


def alerts_for(
    alerts: Iterable[ServiceAlert], stop_id: str, route_ids: frozenset[str]
) -> list[ServiceAlert]:
    """Select alerts naming this stop or any of the given routes."""
    return [
        alert
        for alert in alerts
        if stop_id in alert.informed_stop_ids
        or any(route_id in route_ids for route_id in alert.informed_route_ids)
    ]

File: src/transit.ts
/**
 * Domain layer: joins the realtime feeds to the static schedule and resolves
 * the human-friendly queries the tools accept ("9", "Blue Line", "CTC").
 */

import type { Mode, Route, Schedule, Stop } from './static-gtfs.ts';
import type { ServiceAlert, TripUpdate, VehiclePosition } from './realtime.ts';

export interface ResolvedRouteRef {
  readonly routeId: string;
  readonly name: string;
  readonly longName: string;
  readonly mode: Mode;
}

export interface StopRef {
  readonly stopId: string;
  readonly name: string;
  readonly latitude: number;
  readonly longitude: number;
}

export interface VehicleView {
  /** Vehicle number shown on the bus or train. */
  readonly vehicle: string;
  readonly mode: Mode | 'unknown';
  readonly route: ResolvedRouteRef | undefined;
  readonly headsign: string | undefined;
  readonly tripId: string | undefined;
  readonly latitude: number;
  readonly longitude: number;
  readonly bearingDegrees: number | undefined;
  readonly speedMph: number | undefined;
  readonly occupancy: string | undefined;
  readonly reportedAt: string | undefined;
  readonly positionAgeSeconds: number | undefined;
  readonly nextStop: NextStop | undefined;
}

export interface NextStop {
  readonly stopId: string;
  readonly name: string | undefined;
  readonly arrivalTime: string;
  readonly minutesAway: number;
}

export interface ArrivalView {
  readonly route: ResolvedRouteRef | undefined;
  readonly headsign: string | undefined;
  readonly vehicle: string | undefined;
  readonly tripId: string | undefined;
  readonly arrivalTime: string;
  readonly minutesAway: number;
  readonly scheduledArrivalTime: string | undefined;
  /** Positive means running late. Omitted when the feed gives no schedule. */
  readonly scheduleDeviationMinutes: number | undefined;
  readonly vehiclePosition: { readonly latitude: number; readonly longitude: number } | undefined;
}

const METERS_PER_SECOND_TO_MPH = 2.236_936_3;

function normalize(text: string): string {
  return text.trim().toLowerCase().replace(/\s+/g, ' ');
}

function toIsoTime(epochSeconds: number | undefined): string | undefined {
  if (epochSeconds === undefined) return undefined;
  return new Date(epochSeconds * 1000).toISOString();
}

function minutesBetween(epochSeconds: number, nowMs: number): number {
  return Math.round((epochSeconds * 1000 - nowMs) / 60_000);
}

/** Great-circle distance in meters. */
export function distanceMeters(
  fromLat: number,
  fromLon: number,
  toLat: number,
  toLon: number,
): number {
  const earthRadius = 6_371_000;
  const toRadians = (degrees: number): number => (degrees * Math.PI) / 180;
  const deltaLat = toRadians(toLat - fromLat);
  const deltaLon = toRadians(toLon - fromLon);
  const a =
    Math.sin(deltaLat / 2) ** 2 +
    Math.cos(toRadians(fromLat)) * Math.cos(toRadians(toLat)) * Math.sin(deltaLon / 2) ** 2;
  return 2 * earthRadius * Math.asin(Math.min(1, Math.sqrt(a)));
}

function toRouteRef(route: Route | undefined): ResolvedRouteRef | undefined {
  if (route === undefined) return undefined;
  return {
    routeId: route.routeId,
    name: route.shortName,
    longName: route.longName,
    mode: route.mode,
  };
}

/**
 * Resolves a user-supplied route query against the schedule.
 *
 * Matching is exact-first (`route_id`, then `route_short_name`) so a query of
 * "5" cannot be captured by "Route 501"; substring matching on the long name
 * is the last resort.
 */
export function findRoutes(schedule: Schedule, query: string): Route[] {
  const needle = normalize(query);
  if (needle === '') return [];
  const all = [...schedule.routes.values()];

  const byId = all.filter((route) => normalize(route.routeId) === needle);
  if (byId.length > 0) return byId;

  const byShortName = all.filter((route) => normalize(route.shortName) === needle);
  if (byShortName.length > 0) return byShortName;

  const byLongName = all.filter((route) => normalize(route.longName) === needle);
  if (byLongName.length > 0) return byLongName;

  return all.filter(
    (route) =>
      normalize(route.longName).includes(needle) || normalize(route.shortName).includes(needle),
  );
}

/**
 * Resolves a user-supplied stop query: exact `stop_id`/`stop_code` first, then
 * exact name, then a substring match ranked by name length so the closest
 * match to the query leads.
 */
export function findStops(schedule: Schedule, query: string, limit = 10): Stop[] {
  const needle = normalize(query);
  if (needle === '') return [];

  const direct = schedule.stops.get(query.trim());
  if (direct !== undefined) return [direct];

  const all = [...schedule.stops.values()];
  const byCode = all.filter((stop) => stop.code !== undefined && normalize(stop.code) === needle);
  if (byCode.length > 0) return byCode.slice(0, limit);

  const byName = all.filter((stop) => normalize(stop.name) === needle);
  if (byName.length > 0) return byName.slice(0, limit);

  return all
    .filter((stop) => normalize(stop.name).includes(needle))
    .sort((a, b) => a.name.length - b.name.length || a.name.localeCompare(b.name))
    .slice(0, limit);
}

/** Matches a vehicle query against its label, descriptor id, or entity id. */
export function matchesVehicleQuery(vehicle: VehiclePosition, query: string): boolean {
  const needle = normalize(query);
  if (needle === '') return false;
  const candidates = [vehicle.vehicleLabel, vehicle.vehicleId, vehicle.entityId];
  return candidates.some((value) => value !== undefined && normalize(value) === needle);
}

/**
 * Index of the soonest upcoming stop per trip, built from the TripUpdates feed.
 *
 * This is the only trustworthy source of a vehicle's next stop: the
 * VehiclePositions feed's own `stop_id` and `current_stop_sequence` do not
 * match the published schedule (see the note in `realtime.ts`).
 */
function buildNextStopIndex(
  tripUpdates: readonly TripUpdate[],
  nowMs: number,
): Map<string, { stopId: string; arrivalTime: number }> {
  const index = new Map<string, { stopId: string; arrivalTime: number }>();
  const nowSeconds = nowMs / 1000;

  for (const update of tripUpdates) {
    const tripId = update.tripId;
    if (tripId === undefined) continue;
    let best: { stopId: string; arrivalTime: number } | undefined;
    for (const stopTime of update.stopTimeUpdates) {
      const { stopId, arrivalTime } = stopTime;
      if (stopId === undefined || arrivalTime === undefined) continue;
      if (arrivalTime < nowSeconds) continue;
      if (best === undefined || arrivalTime < best.arrivalTime) {
        best = { stopId, arrivalTime };
      }
    }
    if (best !== undefined) index.set(tripId, best);
  }
  return index;
}

export interface TransitSnapshot {
  readonly schedule: Schedule;
  readonly vehicles: readonly VehiclePosition[];
  readonly tripUpdates: readonly TripUpdate[];
  readonly nowMs: number;
}

/** Builds the enriched, rider-facing view of a single vehicle. */
export function toVehicleView(
  vehicle: VehiclePosition,
  snapshot: TransitSnapshot,
  nextStopIndex: Map<string, { stopId: string; arrivalTime: number }>,
): VehicleView {
  const { schedule, nowMs } = snapshot;
  const trip = vehicle.tripId === undefined ? undefined : schedule.trips.get(vehicle.tripId);
  // Prefer the route the vehicle reports; fall back to the one on its trip.
  const routeId = vehicle.routeId ?? trip?.routeId;
  const route = routeId === undefined ? undefined : schedule.routes.get(routeId);

  const upcoming = vehicle.tripId === undefined ? undefined : nextStopIndex.get(vehicle.tripId);
  const nextStop: NextStop | undefined =
    upcoming === undefined
      ? undefined
      : {
          stopId: upcoming.stopId,
          name: schedule.stops.get(upcoming.stopId)?.name,
          arrivalTime: new Date(upcoming.arrivalTime * 1000).toISOString(),
          minutesAway: minutesBetween(upcoming.arrivalTime, nowMs),
        };

  return {
    vehicle: vehicle.vehicleLabel ?? vehicle.vehicleId ?? vehicle.entityId,
    mode: route?.mode ?? 'unknown',
    route: toRouteRef(route),
    headsign: trip?.headsign,
    tripId: vehicle.tripId,
    latitude: vehicle.latitude,
    longitude: vehicle.longitude,
    bearingDegrees: vehicle.bearingDegrees,
    speedMph:
      vehicle.speedMetersPerSecond === undefined
        ? undefined
        : Math.round(vehicle.speedMetersPerSecond * METERS_PER_SECOND_TO_MPH * 10) / 10,
    occupancy: vehicle.occupancyStatus,
    reportedAt: toIsoTime(vehicle.timestamp),
    positionAgeSeconds:
      vehicle.timestamp === undefined
        ? undefined
        : Math.max(0, Math.round(nowMs / 1000 - vehicle.timestamp)),
    nextStop,
  };
}

export function toVehicleViews(
  vehicles: readonly VehiclePosition[],
  snapshot: TransitSnapshot,
): VehicleView[] {
  const index = buildNextStopIndex(snapshot.tripUpdates, snapshot.nowMs);
  return vehicles.map((vehicle) => toVehicleView(vehicle, snapshot, index));
}

/** Predicted arrivals at one stop, soonest first. */
export function arrivalsAtStop(
  stop: Stop,
  snapshot: TransitSnapshot,
  options: { readonly routeIds?: ReadonlySet<string>; readonly mode?: Mode } = {},
): ArrivalView[] {
  const { schedule, tripUpdates, vehicles, nowMs } = snapshot;
  const nowSeconds = nowMs / 1000;

  const vehicleByTrip = new Map<string, VehiclePosition>();
  for (const vehicle of vehicles) {
    if (vehicle.tripId !== undefined) vehicleByTrip.set(vehicle.tripId, vehicle);
  }

  const arrivals: ArrivalView[] = [];
  for (const update of tripUpdates) {
    const trip = update.tripId === undefined ? undefined : schedule.trips.get(update.tripId);
    const routeId = update.routeId ?? trip?.routeId;
    const route = routeId === undefined ? undefined : schedule.routes.get(routeId);

    if (options.routeIds !== undefined && (routeId === undefined || !options.routeIds.has(routeId)))
      continue;
    if (options.mode !== undefined && route?.mode !== options.mode) continue;

    for (const stopTime of update.stopTimeUpdates) {
      if (stopTime.stopId !== stop.stopId) continue;
      if (stopTime.scheduleRelationship === 'SKIPPED') continue;
      const arrivalTime = stopTime.arrivalTime ?? stopTime.departureTime;
      if (arrivalTime === undefined || arrivalTime < nowSeconds) continue;

      const vehicle =
        update.tripId === undefined ? undefined : vehicleByTrip.get(update.tripId);
      const scheduled = stopTime.scheduledArrivalTime;

      arrivals.push({
        route: toRouteRef(route),
        headsign: trip?.headsign,
        vehicle: vehicle?.vehicleLabel ?? update.vehicleLabel,
        tripId: update.tripId,
        arrivalTime: new Date(arrivalTime * 1000).toISOString(),
        minutesAway: minutesBetween(arrivalTime, nowMs),
        scheduledArrivalTime: toIsoTime(scheduled),
        // The feed omits `delay`, so deviation is derived from the two stamps.
        scheduleDeviationMinutes:
          scheduled === undefined ? undefined : Math.round((arrivalTime - scheduled) / 60),
        vehiclePosition:
          vehicle === undefined
            ? undefined
            : { latitude: vehicle.latitude, longitude: vehicle.longitude },
      });
    }
  }

  return arrivals.sort((a, b) => a.arrivalTime.localeCompare(b.arrivalTime));
}

/** Alerts naming this stop or any of the given routes. */
export function alertsFor(
  alerts: readonly ServiceAlert[],
  stopId: string,
  routeIds: ReadonlySet<string>,
): ServiceAlert[] {
  return alerts.filter(
    (alert) =>
      alert.informedStopIds.includes(stopId) ||
      alert.informedRouteIds.some((routeId) => routeIds.has(routeId)),
  );
}

export { normalize };
